src/System.Management.Automation/engine/parser/PSType.cs

```diff
@@ -542,7 +542,15 @@ public void DefineMembers()
 
             private void DefineProperty(PropertyMemberAst propertyMemberAst)
             {
-                if (_definedProperties.ContainsKey(propertyMemberAst.Name))
+                // Report an error if the property name collides with an existing property or a non-constructor method definition
+                // Property naming collisions with constructors are fine though, "$this.C" unambiguously refers to the property in this example:
+                //
+                //     class C {
+                //       C() {}
+                //       $C
+                //     }
+                //
+                if (_definedProperties.ContainsKey(propertyMemberAst.Name) || (_definedMethods.ContainsKey(propertyMemberAst.Name) && !_definedMethods[propertyMemberAst.Name][0].Item1.IsConstructor))
```

Comment:
nit: Can you please define a local variable to hold `propertyMemberAst.Name` and replace `propertyMemberAst.Name` with that local variable in this method?
That will make this check less cumbersome and a bit more readable.